### PR TITLE
[feature/otx]: Supervised Contrastive Loss for Multiclass classification

### PR DESCRIPTION
**Pull request by arrufat**
This PR contains add the possibility to train a classifier using a custom model with a dual head:
- Classification head: standard
- SelfSL head: currently it uses the BarlowTwins loss, but could be used with other Self-SL losses.

Training in this way provides significant gains in the low data regime (10 samples per class).
In my experiments, the classification error is reduced by about 5% on average.

PS: This is my first PR, so I expect a lot of feedback.

### CODE REVIEW

File: train_cls.py
Comment (sovrasov) [inline train_cls.py]: This file should be deleted prior to merging.

File: otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/model_selfsl.py
Comment (sovrasov) [inline otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/model_selfsl.py]: Dead code

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_selfsl.yaml
Comment (sovrasov) [inline otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_selfsl.yaml]: `train_type` option could be specified via otx CLI, so we just don't need one more copy of template with a different `train_type`.

Overall, training paradigm choice should be done via the `train_type` parameter (supervised, semu-sl, self-sl). Also, we have a subdivision into multiclass, multilabel and hierarchical methods. It's done in runtime and based on the input label schema (and it can't be specified via template in advance). This scheme is not perfect, but at least for now we try to keep only one template per classification model and also this is a requirement from the geti's architecture side.

File: otx/algorithms/classification/tasks/inference.py
Comment (sovrasov) [inline otx/algorithms/classification/tasks/inference.py]: This looks like a hack. The selfsl recipe should be located in the same folder as the other ones (inside the MPA repo).

File: otx/algorithms/classification/tasks/inference.py
Comment (sovrasov) [inline otx/algorithms/classification/tasks/inference.py]: This if could be the outer one:
```
if self._hyperparams.algo_backend.train_type == TrainType.SELFSUPERVISED:
   ...
else:
   base_dir = os.path.abspath(os.path.dirname(self.template_file_path))
   ...
```
Such structure is a bit more clear to extend in the future.

File: otx/algorithms/classification/adapters/mmcls/model/heads/selfsl_cls_head.py
Comment (sovrasov) [inline otx/algorithms/classification/adapters/mmcls/model/heads/selfsl_cls_head.py]: Right now we keep all the heads, losses, etc in the MPA repo. In the future this repo will be moved to otx anyway, but for now I'd keep all the custom modules together in one place.

**Comment by sovrasov** [comment]
Overall, for me the self-sl workflow is not clear. Normally we launch training via otx train. This tool loads the template, creates a task object, dataset, then launches a training. What is the workflow for self-sl classification? I assume, this scenario consumes unlabeled data and then train() method of the classification task produces a feature extractor, saves it's weights.
After that we can launch a supervised training via otx train, and finetune the obtained pre-trained weights for multiclass / multilabel / hierarchical classification on a labelled dataset.
@arrufat as fat as I see, you initially intended to implement a different workflow, may be I didn't fully understand the requirements to the self-sl method, could you please share your vision?

**Review by jaegukhyun — CHANGES_REQUESTED** [review]
I think this PR should be checked by pre-commit hook.
I recommend to use command 'tox -re pre-commit' to check pre-commit issues

**Comment by arrufat** [comment]
> Overall, for me the self-sl workflow is not clear. Normally we launch training via otx train. This tool loads the template, creates a task object, dataset, then launches a training. What is the workflow for self-sl classification? I assume, this scenario consumes unlabeled data and then train() method of the classification task produces a feature extractor, saves it's weights. After that we can launch a supervised training via otx train, and finetune the obtained pre-trained weights for multiclass / multilabel / hierarchical classification on a labelled dataset. @arrufat as fat as I see, you initially intended to implement a different workflow, may be I didn't fully understand the requirements to the self-sl method, could you please share your vision?

Thank you for the feedback, I will address the issues you pointed out soon.

I think I was not clear about the motivation of this PR.
The goal is to improve classification accuracy with resource constraints: few samples per class and few epochs.

The main idea is drawn from the [Supervised Contrastive Learning Paper](https://arxiv.org/abs/2004.11362).
In that paper, they follow the scheme that you described:
- train the model using Supervised contrastive learning (contrastive loss with labels)
- save the weights.
- linear probe the model.

We decided that, if we could combine all of this in one step, we could use the SupCon loss as a guidance for the Cross-Entropy loss. That could lead to faster convergence and higher accuracy in this scenario:
- train a network with 2 heads
  - a head with classification loss (using the labels)
  - a head with SupCon loss (also using the labels)
- after training, discard the head with the SupCon loss

In my experiments, I tried the SupCon approach, and also experimented with different SSL methods. Replacing the SupCon loss with the [Barlow Twins loss](https://arxiv.org/abs/2103.03230) lead to the most significant and consistent gains across datasets and configurations. The thing is, due to the way the Barlow Twins loss works, we're not using the labels for that head.

I agree that the name of the method (SelfSL Classifier) might not be the best I could've come up with, but that's what it is: a combination of a classification + SelfSL method. I am open to suggestions, and maybe I should not use the `TrainType.SELFSUPERVISED` for this task, either.

**Comment by sovrasov** [comment]
Indeed, `TrainType.SELFSUPERVISED` is not the right option for this method. So it looks like we could integrate it into the existing supervised learning recipe then. I'd propose making an additional parameter in classification configurable parameters (see training_extensions/external/model-preparation-algorithm/mpa_tasks/apis/classification/config.py + configuration.yaml). It will be visible in the geti UI, but not from the first glance, and even not from the second... At the same time, I guess we can't just enable this method by default. Introducing dataset size threshold parameter is also an option (SupCon is always on if we have less than x data).

**Comment by kprokofi** [comment]
I agree with @sovrasov 
We don't need to add a new training type. Self-supervised it is something different that we intend to add as a feature in the next Q1-Q2
We can extend existent incremental learning training type.
We may add necessary functionality like losses, heads into MPA and add additional parameters in yaml config for multiclass classification. It will be good to introduce switching on / off rule for this method too.

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_selfsl.yaml
Comment (arrufat) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_selfsl.yaml]: I couldn't find how to specify the `train_type` in the docs. Would you mind showing me how to do it?

Comment (sovrasov) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_selfsl.yaml]: All the hyperparameters in template are exported to command line. The mechanism is described here:  https://github.com/openvinotoolkit/training_extensions/blob/develop/QUICK_START_GUIDE.md#model-template-specific-parameters 
But I just found out that this parameter is marked as non editable, also it has a wrong description. To adjust it via CLI we have to make some modifications in configuration.yaml, but it's a subject of discussion. Sorry for confusion, for now you can add additional templates.

Comment (arrufat) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_selfsl.yaml]: Thank you :)

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Thanks for the new feature.
But we don't think this is a new train type, rather it's an optional accuracy improvement for multi-class classification. In experimental results, it seems it could even replace default supervised one.
So I'd like to suggest that integrate this feature as 'optional enablement' via configurable parameters. You can refer to `enable_tilng` h-param in https://github.com/openvinotoolkit/training_extensions/blob/46b831bcdb7db8e3dc08e819c37969de55c53b4d/external/mmdetection/detection_tasks/apis/detection/configuration.py#L183

File: otx/algorithms/classification/utils/data.py
Comment (jaegukhyun) [inline otx/algorithms/classification/utils/data.py]: Could you explain why this function should be changed?

File: otx/algorithms/classification/configs/configuration.yaml
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/configuration.yaml]: As Songki said, we would like to recommend to add supcon option in configuration.yaml like tiling method in detection

File: otx/algorithms/classification/utils/data.py
Comment (arrufat) [reply otx/algorithms/classification/utils/data.py]: Ah, I can revert this change. This allows loading multiclass datasets from JSON files, and not only from a directory tree. It was useful to subsample the datasets in my experiments. I will revert it, sorry.

**Review by jaegukhyun — CHANGES_REQUESTED** [review]
I left two commend. First is same with Songki's request. Second is about the reason why loading dataset part is changed

**Comment by arrufat** [comment]
@goodsong81 @jaegukhyun how does the last change look like?

**Comment by jaegukhyun** [comment]
@arrufat Let me take a look, Thank you for your effort!!

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_supcon.yaml
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template_supcon.yaml]: This is same with template.yaml. OTX TF have agreed that one template.yaml per one model. You can generate templates for your experiments, but that templates can be pushed.

File: otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/template_supcon.yaml
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/template_supcon.yaml]: As I mentioned, you should remove template_*.yaml

File: otx/algorithms/classification/configs/mobilenet_v3_large_075_cls_incr/template_supcon.yaml
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/mobilenet_v3_large_075_cls_incr/template_supcon.yaml]: Same above

File: otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/template_supcon.yaml
Comment (arrufat) [reply otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/template_supcon.yaml]: Ah, that parameter can be set via the command line?

File: otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/template_supcon.yaml
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/template_supcon.yaml]: Same above

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/model_supcon.py
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/efficientnet_b0_cls_incr/model_supcon.py]: This is optional, how about "EfficientNet-B0 config for multi-class with contrastive loss for smaller dataset"

Comment (arrufat) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/model_supcon.py]: Yes, that sounds better, I will update the descriptions of all models :)

File: otx/algorithms/classification/tasks/inference.py
Comment (jaegukhyun) [inline otx/algorithms/classification/tasks/inference.py:330]: Is there model_supcon_multilabel.py in template directory?

File: otx/algorithms/classification/tasks/inference.py
Comment (jaegukhyun) [inline otx/algorithms/classification/tasks/inference.py]: Is there model_supcon_multilabel.py in template dir?

File: otx/algorithms/classification/tasks/inference.py
Comment (jaegukhyun) [inline otx/algorithms/classification/tasks/inference.py]: If supcon only supports incremental learning(fine-tuning), this blocks should be inside if train_type == TrainType.INCREMENTAL. Because semi-sl and self-sl will be joined

**Review by jaegukhyun — CHANGES_REQUESTED** [review]
I left some comments, I think we are so close to add this feature!!

File: otx/algorithms/classification/tasks/inference.py
Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: There will be soon, once I push the changes in the MPA.
I think this PR should be on hold until the MPA changes are merged.
Then I will make another PR in the MPA adding support for multilabel and hierarchical classification.

Or, if you prefer, I can remove this and make another PR here later.

File: otx/algorithms/classification/tasks/inference.py
Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: Right...

File: otx/algorithms/classification/tasks/inference.py
Comment (jaegukhyun) [reply otx/algorithms/classification/tasks/inference.py]: Yes, let's make another PR for expand SupCon to multilabel and hierarchical classification

**Review by jaegukhyun — DISMISSED** [review]
@sovrasov @goodsong81 It looks good to me. I want you to review this PR once.
@arrufat Is there any plan for test SupCon feature?

**Comment by arrufat** [comment]
> @sovrasov @goodsong81 It looks good to me. I want you to review this PR once. @arrufat Is there any plan for test SupCon feature?

Yes, I need to add integration tests, I just haven't got there yet.

File: otx/algorithms/classification/tasks/inference.py
Comment (sungmanc) [inline otx/algorithms/classification/tasks/inference.py]: How about using one additional `elif` sentence for `supcon` to unify the codes?

AS IS
```
if self._hyperparams.learning_parameters.enable_supcon:
            model_base_name = "model_supcon"

        if self._multilabel:
            cfg_path = os.path.join(base_dir, "model_multilabel.py")
        elif self._hierarchical:
            cfg_path = os.path.join(base_dir, "model_hierarchical.py")
        else:
            cfg_path = os.path.join(base_dir, model_base_name + ".py")
``` 

TO BE
```
        if self._multilabel:
            cfg_path = os.path.join(base_dir, "model_multilabel.py")
        elif self._hierarchical:
            cfg_path = os.path.join(base_dir, "model_hierarchical.py")
        elif self._hyperparameters.learning_parameters.enable_supcon:
            cfg_path = os.path.join(base_dir, "model_supcon.py")
        else:
            cfg_path = os.path.join(base_dir, "model.py")
```

Comment (sungmanc) [reply otx/algorithms/classification/tasks/inference.py]: And actually, I want to abbreviate `self._hyperparameters.learning_parameters.enable_supcon` --> `self._supcon` by adding some codes.

File: otx/algorithms/common/tasks/training_base.py
Comment (sungmanc) [inline otx/algorithms/common/tasks/training_base.py]: Why we need `train_data_cfg["classes"]` ? can't we use existing one(`train_data_cfg["data_classes"]`)?

**Review by sungmanc — COMMENTED** [review]
I left some comments, please check them.

File: otx/algorithms/common/tasks/training_base.py
Comment (arrufat) [reply otx/algorithms/common/tasks/training_base.py]: When I modified the data loading part to be able to load images from a JSON file for the multiclass classification case (https://github.com/openvinotoolkit/training_extensions/pull/1406/commits/cf929c3ab77d3791863337e34494a4012f1e8b97), somehow, it complained that it wanted to see the classes using the `classes`, so I just added that.

I need to revert those two changes, since that was only made for me to be able to easily subsample classification datasets without haveing to duplicate the tree structure several times.

File: otx/algorithms/classification/tasks/inference.py
Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: I wrote it like this because the `enable_supcon` toggle is not at the same level (or exclusive) with the other `TrainType`s, but complimentary. So, it seems misleading to me.

Moreover, in an upcoming PR, there will be `model_supcon_multilabel` and `model_supcon_hierarchical`.

Comment (sungmanc) [reply otx/algorithms/classification/tasks/inference.py]: > I wrote it like this because the `enable_supcon` toggle is not at the same level (or exclusive) with the other `TrainType`s, but complimentary. So, it seems misleading to me.
> 
> Moreover, in an upcoming PR, there will be `model_supcon_multilabel` and `model_supcon_hierarchical`.

If so, the name of `model_supcon.py` is weird to me because the goal of `model_****.py`  is to divide the `TrainType`.
Similarly, `model_supcon_multilabel` and `model_supcon_hierarchical` are also weird. As you said, although `supcon` represents the complementary method and `multilabel` represents the `TrainType`, they'll be located the same level.

Comment (sungmanc) [reply otx/algorithms/classification/tasks/inference.py]: However, actually, in my case, I don't have ideas for a better way, now. We need to think about more

Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: In that case, we could remove the underscore: `modelsupcon_****.py`.
Then we would be more explicit about the fact that it's a different model (at least for training) and keep the underscore as a field separator for the `TrainType`.

**Comment by arrufat** [comment]
Updated to latest `feature/otx`.

File: otx/algorithms/classification/tasks/inference.py
Comment (jaegukhyun) [inline otx/algorithms/classification/tasks/inference.py:360]: We don't have supcon.yaml. If you want to change hyperparameters for supcon, you should change template.yaml manually before training experiments.

**Review by jaegukhyun — CHANGES_REQUESTED** [review]
I add a comment which is related with template.yaml name for supcon. Basically we use one template for one model policy. Therefore if you want to use supcon for your experiments, then you should change hyper parameters in template.yaml manually. Let's discuss about automatic hyper-parameter changes for SupCon later.
And one more thing, I want you to check pre-commit error in github action.

File: otx/algorithms/classification/tasks/inference.py
Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py:360]: But we do have `supcon.yaml` already in `feature/otx`:
https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/recipes/stages/classification/supcon.yaml

Comment (jaegukhyun) [reply otx/algorithms/classification/tasks/inference.py:360]: Ah, sorry about that. I confused with template and recipe.

Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py:360]: No worries, it's very confusing.

**Comment by sovrasov** [comment]
> I add a comment which is related with template.yaml name for supcon. Basically we use one template for one model policy. Therefore if you want to use supcon for your experiments, then you should change hyper parameters in template.yaml manually. Let's discuss about automatic hyper-parameter changes for SupCon later. And one more thing, I want you to check pre-commit error in github action.

Since the new supcon parameter in configuration is editable, it should be accessible via OTX CLI and users can modify it without editing template.yaml. Overall, the changes look good to me.

**Review by jaegukhyun — DISMISSED** [review]
I think we can merge this PR, after solving conflicts
@sungmanc @harimkang @supersoob @JihwanEom Can anyone review this PR once more?

**Comment by arrufat** [comment]
@jaegukhyun conflicts resolved :)

**Comment by jaegukhyun** [comment]
There are pylint issues, could you check it?
pylint...................................................................Failed
- hook id: pylint
- exit code: 4

************* Module otx.algorithms.classification.tasks.inference
otx/algorithms/classification/tasks/inference.py:356:16: W0201: Attribute 'task_pipeline_path' defined outside __init__ (attribute-defined-outside-init)
otx/algorithms/classification/tasks/inference.py:363:12: W0201: Attribute 'task_pipeline_path' defined outside __init__ (attribute-defined-outside-init)
otx/algorithms/classification/tasks/inference.py:362:12: W0201: Attribute 'task_model_dir' defined outside __init__ (attribute-defined-outside-init)

File: otx/algorithms/classification/tasks/inference.py
Comment (jaegukhyun) [inline otx/algorithms/classification/tasks/inference.py]: We agreed on variable name for pipeline and model.
It should be self.model_dir, and self.pipeline_path. Please refer segmentation parts. Therefore I want you to change these variable name

**Review by jaegukhyun — COMMENTED** [review]
I left one comment

File: otx/algorithms/classification/tasks/inference.py
Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: I am not sure I follow, I did it like in the selfsl case.

Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: I didn't update the `self.model_dir` to contain `supcon` because it won't work when the task is `multilabel` or `hierarchical`. In my next PR, which adds support for those two tasks, I will do it.

Comment (jaegukhyun) [reply otx/algorithms/classification/tasks/inference.py]: Does selfsl use self.model_dir and pipeline_path?
            self.model_dir = os.path.join(self.model_dir, "selfsl")
            pipeline_path = os.path.join(os.path.dirname(pipeline_path), "selfsl/data_pipeline.py")

Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: yes

Comment (jaegukhyun) [reply otx/algorithms/classification/tasks/inference.py]: Then why supcon generate new variable task_model_dir and task_pipeline_path?

Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: I don't understand your question. Supcon needs its own model and data pipeline.

Comment (jaegukhyun) [reply otx/algorithms/classification/tasks/inference.py]: My point is why we should change variable name from self.model_dir to self.task_model_dir and from self.pipeline_path to self.task_pipeline_path.
Current feature/otx branch already merged Sunchul's cls-selfsl PR and he agree the variable names self.model_dir and self.pipeline_path

Comment (arrufat) [reply otx/algorithms/classification/tasks/inference.py]: Ah, I see it now, it was changed after I merged the first time, I will fix that. I didn't make this names up, they were carried out from a previous version. Sorry for the confusion.

Comment (jaegukhyun) [reply otx/algorithms/classification/tasks/inference.py]: No problem!

**Review by jaegukhyun — DISMISSED** [review]

**Review by supersoob — DISMISSED** [review]

**Review by goodsong81 — DISMISSED** [review]
LGTM.

File: tests/integration/cli/classification/test_classification.py
Comment (jaegukhyun) [inline tests/integration/cli/classification/test_classification.py:105]: I think it's better to test this before test_otx_train because below eval, export test expect the weight comes from normal incremental learning

Comment (arrufat) [reply tests/integration/cli/classification/test_classification.py:105]: Ok, got it.

**Review by jaegukhyun — APPROVED** [review]
LGTM, Thank you for your effort, if you attach your local test, we will check the results.
I think classification test is enough

**Comment by arrufat** [comment]
Test summary without this PR:
image

Test summary after this PR:
image

**Review by supersoob — APPROVED** [review]

**Review by harimkang — APPROVED** [review]
Nice work! LGTM

**Pull request merged**
